yaml.go

```diff
@@ -4,17 +4,60 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
-	"io"
 	"reflect"
 	"strconv"
 
+	jsoniter "github.com/json-iterator/go"
 	"gopkg.in/yaml.v2"
 )
 
+// The json-iterator implementation is based on:
+// https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go
+
+type customNumberExtension struct {
+	jsoniter.DummyExtension
+}
+
+// createJSONIterator returns a jsoniterator API that's configured to be
+// compatible with the encoding/json standard library.
+func createJSONIterator() jsoniter.API {
+	config := jsoniter.Config{
+		EscapeHTML:             true,
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+	}.Froze()
+	// Force jsoniter to decode number to interface{} via int64/float64, if possible.
+	config.RegisterExtension(&customNumberExtension{})
+	return config
+}
+
+// createCaseSensitiveStrictJSONIterator returns a jsoniterator API that's configured to be
+// case-sensitive, but also disallow unknown fields when unmarshalling. It is compatible with
+// the encoding/json standard library.
+func createCaseSensitiveStrictJSONIterator() jsoniter.API {
+	config := jsoniter.Config{
+		EscapeHTML:             true,
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+		CaseSensitive:          true,
+		DisallowUnknownFields:  true,
+	}.Froze()
+	// Force jsoniter to decode number to interface{} via int64/float64, if possible.
+	config.RegisterExtension(&customNumberExtension{})
+	return config
+}
+
+// Private copies of jsoniter to try to shield against possible mutations
+// from outside. Still does not protect from package level jsoniter.Register*() functions - someone calling them
+// in some other library will mess with every usage of the jsoniter library in the whole program.
+// See https://github.com/json-iterator/go/issues/265
+var jsonIterator = createJSONIterator()
+var caseSensitiveStrictJSONIterator = createCaseSensitiveStrictJSONIterator()
+
 // Marshal marshals the object into JSON then converts JSON to YAML and returns the
 // YAML.
 func Marshal(o interface{}) ([]byte, error) {
-	j, err := json.Marshal(o)
+	j, err := caseSensitiveStrictJSONIterator.Marshal(o)
```

Comment:
@smarterclayton, didn't you see performance regressions in json-iter marshaling?

Reply:
Would they really be relevant for YAML encoding? It's not that we do that in critical paths, do we?

Reply:
Also, this doesn't affect all of YAML code paths in the current form.

Reply:
Btw json-iterator isn't fully compatible with the standard library right now, created https://github.com/json-iterator/go/pull/371. PTAL.

Reply:
https://github.com/json-iterator/go/pull/371 has merged.